transport-classes-io_uring/src/main/java/io/netty5/incubator/channel/uring/AbstractIOUringChannel.java

```diff
@@ -0,0 +1,802 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty5.incubator.channel.uring;
+
+import io.netty5.buffer.Buffer;
+import io.netty5.buffer.BufferAllocator;
+import io.netty5.buffer.DefaultBufferAllocators;
+import io.netty5.buffer.StandardAllocationTypes;
+import io.netty5.channel.AbstractChannel;
+import io.netty5.channel.ChannelException;
+import io.netty5.channel.ChannelOption;
+import io.netty5.channel.ChannelShutdownDirection;
+import io.netty5.channel.EventLoop;
+import io.netty5.channel.ReadHandleFactory;
+import io.netty5.channel.WriteHandleFactory;
+import io.netty5.channel.socket.SocketProtocolFamily;
+import io.netty5.channel.unix.Errors;
+import io.netty5.channel.unix.FileDescriptor;
+import io.netty5.channel.unix.UnixChannel;
+import io.netty5.channel.unix.UnixChannelOption;
+import io.netty5.util.Resource;
+import io.netty5.util.collection.LongObjectHashMap;
+import io.netty5.util.collection.LongObjectMap;
+import io.netty5.util.concurrent.Future;
+import io.netty5.util.concurrent.FutureContextListener;
+import io.netty5.util.concurrent.Promise;
+import io.netty5.util.internal.SilentDispose;
+import io.netty5.util.internal.logging.InternalLogger;
+import io.netty5.util.internal.logging.InternalLoggerFactory;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.net.ConnectException;
+import java.net.InetSocketAddress;
+import java.net.NetworkInterface;
+import java.net.NoRouteToHostException;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.nio.channels.UnresolvedAddressException;
+import java.util.concurrent.Executor;
+import java.util.function.ObjLongConsumer;
+
+import static io.netty5.channel.unix.UnixChannelUtil.computeRemoteAddr;
+
+abstract class AbstractIOUringChannel<P extends UnixChannel>
+        extends AbstractChannel<P, SocketAddress, SocketAddress>
+        implements UnixChannel {
+    static final InternalLogger LOGGER = InternalLoggerFactory.getInstance(AbstractIOUringChannel.class);
+    static final FutureContextListener<Buffer, Void> CLOSE_BUFFER = (b, f) -> SilentDispose.dispose(b, LOGGER);
+    private static final int MAX_READ_AHEAD_PACKETS = 8;
+
+    protected final InternalLogger logger;
+    protected final LinuxSocket socket;
+    private final Promise<Executor> prepareClosePromise;
+    private final Runnable pendingRead;
+    private final Runnable rdHubRead;
+
+    protected volatile boolean active;
+    protected volatile SocketAddress local;
+    protected volatile SocketAddress remote;
+
+    protected SubmissionQueue submissionQueue;
+
+    protected final ObjectRing<Object> readsPending;
+    protected final ObjectRing<Object> readsCompleted; // Either 'Failure', or a message (buffer, datagram, ...).
+    protected final LongObjectMap<Object> cancelledReads;
```

Comment:
nit: please group things together related to visibility and if these are final or not.